src/x_js-scope.php

```diff
@@ -8,28 +8,33 @@
  * of patent rights can be found in the PATENTS file in the same directory.
  */
 
-class :x:js-scope extends :x:element implements XHPAwaitable {
-  use XHPAsync;
+use namespace /*HHAST_IGNORE_ERROR[UseStatementWithAs]*/ Facebook\XHP\Core as x;
+use type Facebook\XHP\HTML\script;
+use namespace HH\Lib\Vec;
 
-  protected async function asyncRender(): Awaitable<XHPRoot> {
-    $calls = Vector { };
-    $instances = Vector { };
-    $this->setContext(':x:js-scope/calls', $calls);
-    $this->setContext(':x:js-scope/instances', $instances);
+xhp class x_js_scope extends x\element {
+  <<__Override>>
+  protected async function renderAsync(): Awaitable<x\node> {
+    $calls = new ScriptDataList();
+    $instances = new ScriptDataList();
+    $this->setContext('x_js_scope/calls', $calls);
+    $this->setContext('x_js_scope/instances', $instances);
 
-    $child_waithandles = Vector { };
+    $child_waithandles = vec[];
     foreach ($this->getChildren() as $child) {
-      if ($child instanceof :x:composable-element) {
+      if ($child is x\node) {
         $child->__transferContext($this->getAllContexts());
+        // Can't use Vec\map_async, since we want these Awaitables to start
+        // when the entire tree is x\node only and all contexts have been transferred.
         $child_waithandles[] = (async () ==> await $child->__flushSubtree())();
```

Comment:
I don't have the context (sorry about the pun :p) to understand what exactly needs to happen here but either the comment is inaccurate or the code is incorrect -- calling the async lambda may (and almost certainly will) eagerly execute some part of it, possibly all of it. Unless I'm missing something, this seems equivalent to map_async.

Reply:
Yeah, you're right.

`$children = await Vec\map_async($this->getChildren(), $c ==> $c as x\node->__flushSubtree());` would be identical. A non-awaited awaitable is not necessarily not started. So this has no effect on ordering.

Reply:
Sorry for addressing this 3 months late :shy_laughter:.